c2rust-analyze/src/context.rs

```diff
@@ -41,6 +41,70 @@ bitflags! {
         const OFFSET_SUB = 0x0020;
         /// This pointer can be freed.
         const FREE = 0x0040;
+
+        /// This pointer is non-null.
+        ///
+        /// [`NON_NULL`] is set (or not) when the pointer is created,
+        /// and it flows forward along dataflow edges.
+        ///
+        /// The permission is [`NON_NULL`] rather than `NULL` because we allow dropping permissions in any assignment.
+        /// This means removing a permission from a pointer's [`PermissionSet`]
+        /// must allow the pointer to take on more values, not restrict it to fewer values.
+        ///
+        /// The following should be set to [`NON_NULL`]:
+        /// * the results of [`Rvalue::Ref`] and [`Rvalue::AddressOf`]
+        /// * a dereferent (i.e. [`READ`]` | `[`WRITE`]), as otherwise this would be UB
+        /// * the argument or result of a known function like [`_.offset`] where it is UB to pass it an invalid (null) pointer
```

Comment:
As this permission flows forward, it doesn't make sense to set it on the input to a dereference or to a function call, as there's nowhere for the permission to flow from there.  A more extreme approach, like taking the presence of `*p` to mean that the local `p` must be `NON_NULL`, would not fit well with our current flow-insensitive analysis.  Here's an example:
```
let p: *const u8 = either_null_or_valid();
if !p.is_null() {
    f(*p);
}
```
Since the analysis is flow-insensitive, there is no way for it to represent the idea that "`p` is nullable, but inside the `if` statement, it can be treated as `NON_NULL`".  `p` must either be nullable (for all time) or `NON_NULL` (for all time).  And making `p` `NON_NULL` and turning it into `&u8` means that there must be a panic somewhere in the case where `either_null_or_valid()` returns null, which (while allowed by our definition of correctness) is not what we want.

I think the "dereferent" point should be removed, and the "argument or result" point should be changed to something like "the result of a known function that only returns non-null pointers".

Reply:
@spernsteiner, in general, why is analysis flow-insensitive, or is that just how it currently is?  It seems like it could be significantly more powerful if it were flow-sensitive.  Then it could rewrite the above to:

```rust
let p: Option<*const u8> = either_null_or_valid();
if let Some(p) = p {
    f(*p);
}
```

and make `f` take a `NON_NULL`.  With nullability especially, it seems like we'll lose a ton of lifting possibilities without flow-sensitive analysis, as this is **the** common pattern of null checking in C.

That said, I can make the adjustments here (remove the deferent point and remove the arguments part) for now.  I was just wondering about this and our approach here.

Reply:
Fixed in 72fc9d031f7508d21a494727f9b4b124bbf277a8.